api/src/main/java/com/epam/pipeline/manager/cloud/gcp/GCPInstancePriceService.java

```diff
@@ -20,70 +20,116 @@
-        double memory = new BigDecimal((double) machineType.getMemoryMb() / 1024)
-                .setScale(2, RoundingMode.HALF_EVEN).doubleValue();
-        return InstanceOffer.builder()
-                .termType(ON_DEMAND_TERM_TYPE)
-                .tenancy(SHARED_TENANCY)
-                .productFamily(INSTANCE_PRODUCT_FAMILY)
-                .sku(machineType.getName())
-                .priceListPublishDate(new Date())
-                .currency(CURRENCY)
-                .instanceType(machineType.getName())
-                .pricePerUnit(0.0)
-                .regionId(regionId)
-                .unit(HOURS_UNIT)
-                .volumeType("SSD")
-                .operatingSystem("Linux")
-                .instanceFamily(readFamily(machineType.getName()))
-                .vCPU(machineType.getGuestCpus())
-                .gpu(0)
-                .memory(memory)
-                .build();
+    private List<InstanceOffer> instanceOffers(final GCPRegion region,
+                                               final GCPMachine machine,
+                                               final Set<GCPResourcePrice> prices) {
+        return Arrays.stream(GCPBilling.values())
+                .map(billing -> InstanceOffer.builder()
+                        .termType(termType(billing))
+                        .tenancy(SHARED_TENANCY)
+                        .productFamily(INSTANCE_PRODUCT_FAMILY)
+                        .sku(machine.getName())
+                        .pricePerUnit(getPrice(machine, billing, prices))
+                        .priceListPublishDate(new Date())
+                        .currency(CURRENCY)
+                        .instanceType(machine.getName())
+                        .regionId(region.getId())
+                        .unit(HOURS_UNIT)
+                        .volumeType("SSD")
+                        .operatingSystem("Linux")
+                        .instanceFamily(WordUtils.capitalizeFully(machine.getFamily()))
+                        .vCPU(machine.getCpu())
+                        .gpu(machine.getGpu())
+                        .memory(machine.getRam())
+                        .build())
+                .collect(Collectors.toList());
+    }
+
+    private String termType(final GCPBilling billing) {
+        return billing == GCPBilling.ON_DEMAND ? ON_DEMAND_TERM_TYPE : PREEMPTIBLE_TERM_TYPE;
+    }
+
+    private Double getPrice(final GCPMachine machine,
+                            final GCPBilling billing,
+                            final Set<GCPResourcePrice> prices) {
+        final Map<GCPResourceType, Optional<GCPResourcePrice>> requiredPrices = Arrays.stream(GCPResourceType.values())
+                .filter(type -> type.isRequired(machine))
+                .collect(Collectors.toMap(Function.identity(),
+                    type -> findPrice(prices, type, billing, type.family(machine))));
+        final Optional<GCPResourceType> typeWithMissingPrice = requiredPrices.entrySet()
```

Comment:
The following implementation returns `0` if some of the machine resources (cpu, ram or gpu) have no associated SKUs. If we filter and log then the calculated price will be a sum of only available resources. And It is quite misleading because the calculated price is lower than the actual price.

What do you think?

Reply:
Ok, let's leave it as is.